src/upsample.jl

```diff
@@ -105,151 +205,37 @@ can be directly specified with a keyword argument.
 The size of the output is equal to
 `(scale*S1, S2, S3)`, where `S1, S2, S3 = size(x)`.
 """  # the user facing function
-function upsample_linear(x::AbstractArray{<:Any,N}, scale::NTuple{M,Real}) where {N,M}
+function upsample_linear(x::AbstractArray{<:Any,N}, scale::NTuple{M,Real}; align_corners::Bool = true) where {N,M}
     M == N-2 || error("The scale argument should be an NTuple with length $(N-2), but it has length $M.")
     outsize = ntuple(i -> floor(Int, scale[i] * Base.size(x, i)), N-2)
-    return upsample_linear(x; size=outsize)
+    return upsample_linear(x; size=outsize, align_corners)
 end
 
 # convenience for single-number scale
-upsample_linear(x::AbstractArray{<:Any,N}, scale::Real) where N = upsample_linear(x, ntuple(_ -> scale, N-2))
+upsample_linear(x::AbstractArray{<:Any,N}, scale::Real; align_corners::Bool = true) where N =
+    upsample_linear(x, ntuple(_ -> scale, N-2); align_corners)
 
 # this actually calls the upsamling kernel
-function upsample_linear(x::AbstractArray{T,N}; size::Union{Integer, NTuple{<:Any,Integer}}) where {T,N}
+function upsample_linear(x::AbstractArray{T,N}; size::Union{Integer, NTuple{<:Any,Integer}}, align_corners::Bool = true) where {T,N}
     length(size) == N-2 || error("The scale argument should be an NTuple with length $(N-2), but it has length $(length(size)).")
 
     if Base.size(x)[1:N-2] == size
         return x
     end
 
     y = similar(x, T, size..., Base.size(x)[end-1:end]...)
-    return upsample_linear_kernel!(y, x)
+    return upsample_linear_kernel!(y, x; align_corners)
 end
 
 # Convenience definition for integers. The algo internally works with floats and then rounds.
-function upsample_linear(x::AbstractArray{T,<:Any}; size) where T<:Integer
+function upsample_linear(x::AbstractArray{T,<:Any}; size, align_corners::Bool = true) where T<:Integer
     y = float.(x)
-    res = upsample_linear(y; size=size)
+    res = upsample_linear(y; size=size, align_corners)
     return round.(T, res)
 end
 
-# compatibility layer for old versions of NNlibCUDA
-# old versions overload upsample_linear_wcn, new versions overload upsample_linear_kernel
-# can be removed from NNlib 0.9, i.e. revert https://github.com/FluxML/NNlib.jl/pull/414
-# IF https://github.com/FluxML/NNlibCUDA.jl/pull/49 has been merged
-upsample_linear_kernel!(y::AbstractArray{<:Any,3}, x::AbstractArray{<:Any,3}) = upsample_linear_wcn!(y,x)
-upsample_linear_kernel!(y::AbstractArray{<:Any,4}, x::AbstractArray{<:Any,4}) = upsample_bilinear_whcn!(y,x)
-upsample_linear_kernel!(y::AbstractArray{<:Any,5}, x::AbstractArray{<:Any,5}) = upsample_trilinear_whdcn!(y,x)
-∇upsample_linear_kernel!(y::AbstractArray{<:Any,3}, x::AbstractArray{<:Any,3}) = ∇upsample_linear_wcn!(y,x)
-∇upsample_linear_kernel!(y::AbstractArray{<:Any,4}, x::AbstractArray{<:Any,4}) = ∇upsample_bilinear_whcn!(y,x)
-∇upsample_linear_kernel!(y::AbstractArray{<:Any,5}, x::AbstractArray{<:Any,5}) = ∇upsample_trilinear_whdcn!(y,x)
```

Comment:
Unless the plan is to have this PR be part of NNlib 0.9, I think we at least need aliases from the old function names to the new ones.

Reply:
You mean aliases that would allow non-matching eltypes for `y` & `x`?
NNlibCUDA [has](https://github.com/FluxML/NNlibCUDA.jl/blob/5f797aec23cbb5483788697e9e911685b681bbf8/src/upsample.jl#L57) restriction that allows only matching eltypes.
We can relax it though, if it makes sense (probably makes more for CPU only).

Or aliases for `upsample_linear_wcn!`, etc.?
Which for this PR would look like:
```julia
upsample_linear_wcn!(y, x) = upsample_linear_kernel!(y, x)
```

Reply:
The latter.

Reply:
Done.